nexus/src/app/crucible.rs

```diff
@@ -0,0 +1,1051 @@
+
+/// Provides a way for (with BackoffError) Permanent errors to have a different error type than
+/// Transient errors.
+#[derive(Debug, thiserror::Error)]
+enum WaitError {
+    #[error("Transient error: {0}")]
+    Transient(#[from] anyhow::Error),
+
+    #[error("Permanent error: {0}")]
+    Permanent(#[from] Error),
+}
+
+impl super::Nexus {
+    fn crucible_agent_client_for_dataset(
+        &self,
+        dataset: &db::model::Dataset,
+    ) -> CrucibleAgentClient {
+        CrucibleAgentClient::new_with_client(
+            &format!("http://{}", dataset.address()),
+            self.reqwest_client.clone(),
+        )
+    }
+
+    /// Return if the Crucible agent is expected to be there and answer Nexus:
+    /// true means it's gone, and the caller should bail out of the
+    /// ProgenitorOperationRetry loop.
+    async fn crucible_agent_gone_check(
+        &self,
+        dataset_id: Uuid,
+    ) -> Result<bool, Error> {
+        let on_in_service_physical_disk = self
+            .datastore()
+            .dataset_on_in_service_physical_disk(dataset_id)
+            .await?;
+
+        Ok(!on_in_service_physical_disk)
+    }
+
+    /// Call out to Crucible agent and perform region creation.
+    async fn ensure_region_in_dataset(
+        &self,
+        log: &Logger,
+        dataset: &db::model::Dataset,
+        region: &db::model::Region,
+    ) -> Result<Region, Error> {
+        let client = self.crucible_agent_client_for_dataset(dataset);
+        let dataset_id = dataset.id();
+
+        let Ok(extent_count) = u32::try_from(region.extent_count()) else {
+            return Err(Error::internal_error(
+                "Extent count out of range for a u32",
+            ));
+        };
+
+        let region_request = CreateRegion {
+            block_size: region.block_size().to_bytes(),
+            extent_count,
+            extent_size: region.blocks_per_extent(),
+            // TODO: Can we avoid casting from UUID to string?
+            // NOTE: This'll require updating the crucible agent client.
+            id: RegionId(region.id().to_string()),
+            encrypted: region.encrypted(),
+            cert_pem: None,
+            key_pem: None,
+            root_pem: None,
+            source: None,
+        };
+
+        let create_region = || async {
+            let region = match ProgenitorOperationRetry::new(
+                || async { client.region_create(&region_request).await },
+                || async { self.crucible_agent_gone_check(dataset_id).await },
+            )
+            .run(log)
+            .await
+            {
+                Ok(v) => Ok(v),
+
+                Err(e) => {
+                    error!(
+                        log,
+                        "region_create saw {:?}",
+                        e;
+                        "region_id" => %region.id(),
+                        "dataset_id" => %dataset_id,
+                    );
+
+                    match e {
```

Comment:
It's really a bee in my bonnet that we have to do these matches every time, but I"m not sure how to avoid it.

Reply:
See ff3915033 for some reuse :)

Reply:
That is so much better. Thank you!